### PR TITLE
Add AnimatedChildSwitcher widget

### PR DESCRIPTION
**Pull request by gspencergoog**
The AutoFade widget (originally authored by Hixie), will cross-fade between a new child, and a previous child (or children, if the switch happens faster than the fade finishes).

It's a good candidate for places where a widget will be added/removed from a slot, and you want a nice transition to occur.

### CODE REVIEW

**Comment by gspencergoog** [comment]
Is AutoFade a good name for this?  It is somewhat descriptive, which is why I kept it, but what about something like AnimatedChildSwitcher or something?  AutoFade just seems a little too general for what this actually does.

**Comment by lukaspili** [comment]
+1 for ***Switcher. it's explicit and it's also a common term in Android sdk: https://developer.android.com/reference/android/widget/ViewSwitcher.html

There are no such concept in iOS. The closer to this is a UIView animation between two views, with a `transitionCrossDissolve`.

**Comment by gspencergoog** [comment]
OK, done!

File: packages/flutter/lib/src/widgets/animated_child_switcher.dart [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+/// one is fading in if they are swapped fast enough (i.e. before [duration]
+/// elapses).
+///
+/// See also:
+///
+///  * [AnimatedCrossFade], which only fades between two children, but also
+///    interpolates their sizes, and is reversible.
+///  * [FadeTransition] which [AnimatedChildSwitcher] uses to perform the transition.
+class AnimatedChildSwitcher extends StatefulWidget {
+  const AnimatedChildSwitcher({
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/animated_child_switcher.dart]: Constructor doc: which parameters must be non-null

**Review by HansMuller — APPROVED** [review]
LGTM

**Comment by HansMuller** [comment]
[discussed this with @gspencergoog] This widget is nearly generic wrt to the actual animated transition. If the fadeIn fadeOut parameters were renamed and perhaps the alignment parameter was left out, an update could add a builder parameter that would enable building other kinds of transitions.

File: packages/flutter/lib/src/widgets/animated_child_switcher.dart (hunk 1 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/widgets/animated_child_switcher.dart]: Done.

**Comment by Hixie** [comment]
follow-ups in https://github.com/flutter/flutter/pull/16250

**Pull request merged**